plugins/salesforce/core/src/main/java/org/pentaho/di/trans/steps/salesforceinput/SalesforceInput.java

```diff
@@ -447,4 +462,160 @@ public void dispose( StepMetaInterface smi, StepDataInterface sdi ) {
     }
     super.dispose( smi, sdi );
   }
+
+
+  @SuppressWarnings( "java:S1144" ) // Using reflection this method is being invoked
+  private JSONObject getFieldsAction() {
+    JSONObject response = new JSONObject();
+    response.put( StepInterface.ACTION_STATUS, StepInterface.FAILURE_RESPONSE );
+    try {
+      FieldsResponse fieldsResponse = getFields();
+      response.put( "fieldsResponse", fieldsResponse );
+      response.put( StepInterface.ACTION_STATUS, StepInterface.SUCCESS_RESPONSE );
+    } catch ( Exception e ) {
+      response.put( "errorMsg", BaseMessages.getString( PKG, "SalesforceInputMeta.ErrorRetrieveData.DialogMessage" ) );
+      response.put( "details", e.getStackTrace() );
+      log.logError( e.getMessage() );
+    }
+    return response;
+  }
+
+  public FieldsResponse getFields() throws Exception {
+    SalesforceConnection connection = null;
+    try {
+      SalesforceInputMeta meta = (SalesforceInputMeta) getStepMetaInterface();
+      String realURL = getTransMeta().environmentSubstitute( meta.getTargetURL() );
+      String realUsername = getTransMeta().environmentSubstitute( meta.getUsername() );
+      String realPassword = Utils.resolvePassword( getTransMeta(), meta.getPassword() );
+      int realTimeOut = Const.toInt( getTransMeta().environmentSubstitute( meta.getTimeout() ), 0 );
+
+      connection = new SalesforceConnection( log, realURL, realUsername, realPassword );
+      connection.setTimeOut( realTimeOut );
+      FieldsResponse fieldsResponse = new FieldsResponse( new ArrayList<>(), new HashSet<>() );
+      if ( meta.isSpecifyQuery() ) {
+        // Free hand SOQL
+        String realQuery = getTransMeta().environmentSubstitute( meta.getQuery() );
+        connection.setSQL( realQuery );
+        connection.connect();
+        // We are connected, so let's query
+        XmlObject[] fields = connection.getElements();
+        int nrFields = fields.length;
+        for ( int i = 0; i < nrFields; i++ ) {
+          fieldsResponse.fieldDTOList.addAll( addFields( "", fieldsResponse.fieldNames, fields[ i ] ) );
+        }
+      } else {
+        connection.connect();
+
+        Field[] fields = connection.getObjectFields( meta.getModule() );
+        for ( int i = 0; i < fields.length; i++ ) {
+          Field field = fields[ i ];
+          fieldsResponse.fieldNames.add( field.getName() );
+          fieldsResponse.fieldDTOList.add( addField( field ) );
+        }
+      }
+      return fieldsResponse;
+
+    } finally {
+      if ( connection != null ) {
+        try {
+          connection.close();
+        } catch ( Exception e ) {
+          // Ignore errors
+        }
+      }
+    }
+
+  }
+
+  private boolean isNullIdField( XmlObject field ) {
+    return field.getName().getLocalPart().equalsIgnoreCase( "ID" ) && field.getValue() == null;
```

Comment:
can this be changed Objects.nonNull(field.getName()  && "ID" .equalsIgnoreCase( field.getName().getLocalPart() ))
to avoid exception. It may be existing code, minor changes we can still do I believe to avoid exceptions in kettle/webttle.

Reply:
Addressed